internal/pkg/agent/transpiler/rules.go

```diff
@@ -669,6 +669,42 @@ func (r *InjectStreamProcessorRule) Apply(_ AgentInfo, ast *AST) (err error) {
 		namespace := datastreamNamespaceFromInputNode(inputNode)
 		datastreamType := datastreamTypeFromInputNode(inputNode, r.Type)
 
+		var inputID *StrVal
+		inputIDNode, found := inputNode.Find("id")
+		if found {
+			inputID, _ = inputIDNode.Value().(*StrVal)
+		}
+
+		if inputID != nil {
+			// get input-level processors node
+			processorsNode, found := inputNode.Find("processors")
+			if !found {
+				processorsNode = &Key{
+					name:  "processors",
+					value: &List{value: make([]Node, 0)},
+				}
+
+				inputMap, ok := inputNode.(*Dict)
+				if ok {
+					inputMap.value = append(inputMap.value, processorsNode)
+				}
+			}
+
+			processorsList, ok := processorsNode.Value().(*List)
+			if !ok {
+				return errors.New("InjectStreamProcessorRule: input processors is not a list")
+			}
+
+			// inject `input_id` on the input level
+			processorMap := &Dict{value: make([]Node, 0)}
+			processorMap.value = append(processorMap.value, &Key{name: "target", value: &StrVal{value: "source"}})
```

Comment:
I think you are currently adding these fields into the event payload, such that they will be sent to and stored in Elasticsearch eventually. 

This change makes every event sent from agent slightly larger, which I don't think we want. For now they will exclusively be used by the shipper and nothing else.

Let's include the input ID and stream ID in the event metadata instead. Then the gRPC output should be able to extract them and add them to the relevant protobuf message fields without needing to store them in Elasticsearch afterwards.

The `add_fields` documentation explains how to do this https://www.elastic.co/guide/en/beats/filebeat/current/add-fields.html:

```yml
processors:
  - add_fields:
      target: '@metadata'
      fields:
        # ...
```

Reply:
@cmacknz I actually worked on supporting `@metadata` as the `target` parameter in `adds_fields` processor, at last it showed itself useful :)

I've updated everything accordingly.